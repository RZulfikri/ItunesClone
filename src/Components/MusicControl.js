import React, {PureComponent} from 'react';
import {Animated, StyleSheet, TouchableOpacity, View} from 'react-native';
import Colors from '../Themes/Colors';
import Icons from 'react-native-vector-icons/MaterialCommunityIcons';
import TrackPlayer, {STATE_PAUSED} from 'react-native-track-player';
import ProgressBar from './ProgressBar';
import ActiveSongItem from './ActiveSongItem';
import ItunesActions from '../Redux/ItunesRedux';
import {connect} from 'react-redux';

const MAX_HEIGHT = 100;
const MIN_HEIGHT = 0;

const styles = StyleSheet.create({
  container: {
    borderTopWidth: 1,
    borderTopColor: Colors.border,
    paddingHorizontal: 15,
    overflow: 'hidden',
    justifyContent: 'center',
  },
  topContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
  },
  actionContainer: {
    width: 50,
    height: 40,
  },
  playActionContainer: {marginHorizontal: 10},
  bottomContainer: {},
  infoContainer: {
    flex: 1,
  },
  playerContainer: {
    flex: 1,
    flexDirection: 'row',
  },
});

/**
 * This component is used as music player/control.
 * the feature is play, pause, stop, seek, backward and forward music/song
 */

class MusicControl extends PureComponent {
  activeSong;

  constructor(props) {
    super(props);
    this.state = {
      isPause: false,
      isStop: false,
      containerHeight: new Animated.Value(0),
    };

    this.show = this.show.bind(this);
    this.hide = this.hide.bind(this);
    this.play = this.play.bind(this);
    this.pause = this.pause.bind(this);
    this.stop = this.stop.bind(this);
    this.backward = this.backward.bind(this);
    this.forward = this.forward.bind(this);
    this.seek = this.seek.bind(this);
  }

  async componentDidMount() {
    const {setRef} = this.props;

    // action to passing function from child to parent component
    if (typeof setRef === 'function') {
      setRef({
        show: this.show,
        hide: this.hide,
        play: this.play,
        pause: this.pause,
        stop: this.stop,
      });
    }

    // an event listener for player playback state
    this.playbackListener = await TrackPlayer.addEventListener(
      'playback-state',
      async ({state}) => {
        if (state === STATE_PAUSED) {
          const position = await TrackPlayer.getPosition();
          const duration = await TrackPlayer.getDuration();
          if (duration > 0 && position > 0) {
            if (position === duration) {
              this.setState({isPause: true, isStop: true});
            } else {
              this.setState({isPause: true, isStop: false});
            }
          }
        }
      },
    );
  }

  async componentWillUnmount() {
    await TrackPlayer.stop();
    await TrackPlayer.destroy();
    this.playbackListener.remove();
  }

  // action to start show animation
  show() {
    Animated.timing(this.state.containerHeight, {
      toValue: MAX_HEIGHT,
      duration: 300,
      useNativeDriver: false,
    }).start(({finished}) => {
      /* completion callback */
      if (finished) {
        this.isShown = true;
      }
    });
  }

  // action to start hide animation
  hide() {
    Animated.timing(this.state.containerHeight, {
      toValue: MIN_HEIGHT,
      duration: 300,
      useNativeDriver: false,
    }).start(({finished}) => {
      /* completion callback */
      if (finished) {
        this.isShow = false;
      }
    });
  }

  // function to play/resume music
  play(data, force) {
    this.activeSong = data;
    if (this.state.isStop || force) {
      this.setState({isPause: false, isStop: false}, async () => {
        await TrackPlayer.stop();
        await TrackPlayer.setupPlayer();

        // Add a track to the queue
        await TrackPlayer.add({
          id: this.activeSong.get('trackId'),
          url: this.activeSong.get('previewUrl'),
          title: this.activeSong.get('trackName'),
          artist: this.activeSong.get('artistName'),
          artwork: this.activeSong.get('artworkUrl60'),
        });

        // Start playing it
        await TrackPlayer.play();
      });
    } else {
      this.setState({isPause: false, isStop: false}, async () => {
        await TrackPlayer.play();
      });
    }
  }

  // function to pause music player
  pause() {
    this.setState({isPause: true}, async () => {
      await TrackPlayer.pause();
    });
  }

  // function to stop music player
  stop() {
    this.setState({isPause: false, isStop: true}, async () => {
      this.activeSong = undefined;
      await TrackPlayer.stop();
    });
  }

  // function to seek playe based on slider position
  async seek(position) {
    this.setState({isPause: false}, async () => {
      const duration = await TrackPlayer.getDuration();
      await TrackPlayer.seekTo(position * duration);
      await TrackPlayer.play();
    });
  }

  // function to backward music
  backward() {
    const {setBackward} = this.props;
    setBackward();
  }

  // function to forward music
  forward() {
    const {setForward} = this.props;
    setForward();
  }

  render() {
    const {isPause, containerHeight} = this.state;
    return (
      <Animated.View style={[styles.container, {height: containerHeight}]}>
        <View style={styles.topContainer}>
          <View style={styles.infoContainer}>
            <ActiveSongItem />
          </View>
          <View style={styles.playerContainer}>
            <TouchableOpacity
              style={styles.actionContainer}
              activeOpacity={0.8}
              onPress={this.backward}>
              <Icons name={'skip-backward'} size={35} />
            </TouchableOpacity>
            <TouchableOpacity
              style={[styles.actionContainer, styles.playActionContainer]}
              activeOpacity={0.8}
              onPress={isPause ? () => this.play(this.activeSong) : this.pause}>
              <Icons name={isPause ? 'play' : 'pause'} size={35} />
            </TouchableOpacity>
            <TouchableOpacity
              style={styles.actionContainer}
              activeOpacity={0.8}
              onPress={this.forward}>
              <Icons name={'skip-forward'} size={35} />
            </TouchableOpacity>
          </View>
        </View>
        <View style={styles.bottomContainer}>
          <ProgressBar onPause={this.pause} onSeek={this.seek} />
        </View>
      </Animated.View>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    setBackward: () => dispatch(ItunesActions.setBackward()),
    setForward: () => dispatch(ItunesActions.setForward()),
  };
};

export default connect(null, mapDispatchToProps)(MusicControl);
