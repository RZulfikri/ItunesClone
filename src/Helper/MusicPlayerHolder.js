class MusicPlayerHolder {
  player = undefined;
  activeSong = undefined;
  isShow = false;

  constructor() {
    this.setInstance = this.setInstance.bind(this);
  }

  setInstance(instance) {
    this.player = instance;
  }

  hide() {
    this.player.hide();
    this.isShow = false;
  }

  play(data) {
    if (!this.isShow) {
      this.player.show();
      this.isShow = true;
    }
    this.activeSong = data;
    this.player.play(data, true);
  }

  pause() {
    this.player.pause();
  }

  stop() {
    this.player.hide();
    this.isShow = false;
    this.activeSong = undefined;
    this.player.stop();
  }
}

export default new MusicPlayerHolder();
