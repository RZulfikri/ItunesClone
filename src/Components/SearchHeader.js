import React, {PureComponent} from 'react';
import {StyleSheet, Text, TextInput, View} from 'react-native';
import debounce from 'debounce';
import Colors from '../Themes/Colors';
import ItunesActions from '../Redux/ItunesRedux';
import {connect} from 'react-redux';
import MusicPlayerHolder from '../Helper/MusicPlayerHolder';

const styles = StyleSheet.create({
  container: {
    padding: 15,
    borderBottomWidth: 1,
    borderBottomColor: Colors.border,
  },
  textInput: {
    borderWidth: 1,
    borderRadius: 17.5,
    height: 35,
    paddingHorizontal: 15,
    borderColor: Colors.border,
  },
});

class SearchHeader extends PureComponent {
  constructor(props) {
    super(props);

    this.onChangeText = debounce(this.onChangeText.bind(this), 300);
  }

  onChangeText(text) {
    const {searchArtistRequest} = this.props;
    MusicPlayerHolder.stop();
    searchArtistRequest(text);
  }

  render() {
    return (
      <View style={styles.container}>
        <TextInput
          placeholder={'Search Artist'}
          onChangeText={this.onChangeText}
          style={styles.textInput}
        />
      </View>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    searchArtistRequest: keywords =>
      dispatch(ItunesActions.searchArtistRequest(keywords)),
  };
};

export default connect(null, mapDispatchToProps)(SearchHeader);
