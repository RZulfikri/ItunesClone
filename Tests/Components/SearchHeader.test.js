import 'react-native';
import React from 'react';
import SearchHeader from '../../src/Components/SearchHeader';
import renderer from 'react-test-renderer';
import {fromJS, Map} from 'immutable';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import MusicPlayerHolder from '../../src/Helper/MusicPlayerHolder';
import MusicControl, {mockPlay} from '../__mocks__/MusicControl';
import ItunesActions from '../../src/Redux/ItunesRedux';
import debounce from 'debounce';

jest.mock('debounce', () => jest.fn(fn => fn));

const mockStore = configureStore([]);

describe('SearchHeader React-Redux Component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore();

    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <SearchHeader />
      </Provider>,
    );
  });

  it('should renders correctly', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
