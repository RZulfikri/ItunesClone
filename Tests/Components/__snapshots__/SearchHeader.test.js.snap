// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchHeader React-Redux Component should renders correctly 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "#E5E5E5",
      "borderBottomWidth": 1,
      "padding": 15,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Search Artist"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "#E5E5E5",
        "borderRadius": 17.5,
        "borderWidth": 1,
        "height": 35,
        "paddingHorizontal": 15,
      }
    }
    underlineColorAndroid="transparent"
  />
</View>
`;
